### ISSUE TITLE
[MDEP-594] Dependency reported as "unused declared" when runtime-retention type annotation is used

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-dependency-plugin#1029]
**[Andreas Hubold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ahubold)** opened **[MDEP-594](https://issues.apache.org/jira/browse/MDEP-594?redirect=false)** and commented

In the attached example project, module "usage" uses annotation `@Example`:

```
public interface Usage {
  @Example String getEmail(); 
}
```

which is defined in module "annotation" as:

```
@Target({ElementType.TYPE_USE, ElementType.TYPE_PARAMETER})
@Retention(RetentionPolicy.RUNTIME)
public @interface Example {
}
```

Consequently, the pom.xml of module "usage" defines a compile-dependency to module "annotation" but the dependency:analyze-only goal incorrectly reports this as unused dependency:

```
[main] [INFO] --- maven-dependency-plugin:3.0.0:analyze-only (analyze-dependencies) @ usage ---
[main] [WARNING] Unused declared dependencies found:
[main] [WARNING]    com.example:annotation:jar:1.0-SNAPSHOT:compile
```

This false positive is only reported if
* the annotation is meta-annotated with target TYPE_USE + TYPE_PARAMETER
* and the source.version/target.version of the compiler plugin for module "usage" is 1.8

There's no false positive if the `@Target` meta-annotation is removed or if compiler's source/target version is set to 1.7.

---

**Affects:** 3.0.0

**Attachments:**
- [example.tgz](https://issues.apache.org/jira/secure/attachment/12903240/example.tgz) (_1.32 kB_)

**Issue Links:**
- [MSHARED-674](https://issues.apache.org/jira/browse/MSHARED-674) Maven Dependency Analyzer ignores Java 8 type annotations
 (_**"is blocked by"**_)
- [MSHARED-792](https://issues.apache.org/jira/browse/MSHARED-792) Maven Dependency Analyzer ignores Java 8 type annotations on local variables

**Comment by jira-importer** [comment]
**[Andreas Hubold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ahubold)** commented

I first thought this would be the same issue as MDEP-417, but in that ticket the used annotation is "javax.persistence.Table" which is meta-annotated with `@Target(value=TYPE)`. I could not reproduce false positives with that target but only with the new Java 8 target values TYPE_USE and TYPE_PARAMETER. So MDEP-417 seems to be a different (maybe related) case.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

The plugin uses the [Maven Dependency Analyzer](https://maven.apache.org/shared/maven-dependency-analyzer/), which uses ASM. Probably easiest to write a unittest there, but it must have to do with the content of the compiled classfile.

**Comment by jira-importer** [comment]
**[Andreas Hubold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ahubold)** commented

It's indeed a bug in the Maven Dependency Analyzer. I've created a ticket for it with attached patch (proposed fix and test case): MSHARED-674

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Hasn't that being fixed with 3.0.1/3.0.2 of maven-dependency-plugin ?

**Comment by jira-importer** [comment]
**[Andreas Hubold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ahubold)** commented

I can confirm that this bug is fixed with 3.1.0 but not in 3.0.x.

Note that type annotations on local variable declarations are still not detected but that's a different issue: MSHARED-792
